yellowbrick/classifier/class_balance_heat_map.py

```diff
@@ -0,0 +1,169 @@
+"""
+
+##########################################################################
+## Imports
+##########################################################################
+
+from matplotlib import cm
+import matplotlib.pyplot as plt
+import numpy as np
+
+from .base import ClassificationScoreVisualizer
+
+from sklearn.model_selection import train_test_split
+from sklearn.utils.multiclass import unique_labels
+
+
+##########################################################################
+## Class Balance Heat Map Chart
+##########################################################################
+
+class ClassBalanceHeatMap(ClassificationScoreVisualizer):
+    """
+    Class balance heat map that shows the support for each class in the
+    fitted classification model displayed as a stacked bar.  Each bar is
+    segmented to show the distribution of predicted classes for each
+    class. It is initialized with a fitted model and generates a
+    class balance heat map chart on draw.
+
+    Parameters
+    ----------
+    ax: axes
+        the axis to plot the figure on.
+    model: estimator
+        Scikit-Learn estimator object. Should be an instance of a classifier,
+        else ``__init__()`` will raise an exception.
+    classes: list
+        A list of class names for the legend. If classes is None and a y value
+        is passed to fit then the classes are selected from the target vector.
+    kwargs: dict
+        Keyword arguments passed to the super class. Here, used
+        to colorize the bars in the histogram.
+    Notes
+    -----
+    These parameters can be influenced later on in the visualization
+    process, but can and should be set as early as possible.
+    """
+
+    def score(self, X, y=None, **kwargs):
+        """
+        Generates a 2D array where each row is the count of the
+        predicted classes and each column is the true class
+
+        Parameters
+        ----------
+        X : ndarray or DataFrame of shape n x m
+            A matrix of n instances with m features
+        y : ndarray or Series of length n
+            An array or series of target or class values
+
+        Returns
+        -------
+
+        ax : the axis with the plotted figure
+        """
+
+        y_pred = self.predict(X)
+        self.indices = unique_labels(y, y_pred)
+        self.classes_ = self.classes_ if self.classes_ is not None else self.indices
+        self.scores = np.array([
+            [(y_pred[y == label_t] == label_p).sum()
+            for label_p in self.indices]
+            for label_t in self.indices
+            ])
+        return self.draw()
+
+    def draw(self):
+        """
+        Renders the class balance heap map across the axis.
+        Returns
+        -------
+        ax : the axis with the plotted figure
+        """
+
+        prev = np.zeros(len(self.classes_))
+        colors = [cm.spectral(x) for x in np.linspace(0, 1, len(self.indices))]
```

Comment:
I had advised you not to store `self.indices` above, and that would definitely cause a problem here! During score you validated the classes in `y_true` and `y_pred` so at this point, indices can simply be `np.arange(len(self.classes_))`, which is easy enough to recompute in the draw function.

Reply:
Additionally, we should allow the user to specify the color palette rather than hardcoding `cm.spectral`. See `resolve_colors` for more on this.

Reply:
I have run into an issue with the colors .  @bbengfort you asked that I not hard code in cmap thus I wrote this snippet(see below) and I ran the code in jupyter notebook (code below).  I have two different outcomes depending on whether or not I specify a color in the visualizer.  If I don't specify a color,  I want to display a default color scheme.  However, I get an IndexError when no color is specified.

``python
```
if self.color is not None:
    colors = resolve_colors(
                  color=self.color,
                  num_colors=len(self.classes_))
else:
    # colors = [cm.spectral(x) for x in np.linspace(0, 1, len(indices))]
    colors = get_color_cycle()
```

digits = load_digits()
X_train, X_test, y_train, y_test = tts(digits.data, digits.target, test_size=0.33, random_state=42)
visualizer = ClassBalanceHeatMap(RandomForestClassifier(), color='bgrmyck')
visualizer.fit(X_train, y_train)
visualizer.score(X_test, y_test)
visualizer.poof()


### Specify Color: BGRMYCK
![heatmap_rgb](https://user-images.githubusercontent.com/1000117/37115814-4241f146-221a-11e8-9917-6308647c8f74.png)

### No Color Specified: Default scheme
![heatmap_indexerror](https://user-images.githubusercontent.com/1000117/37116255-bbddc010-221b-11e8-92aa-4418c81304d1.png)

Reply:
I'm pretty sure you can pass None into `resolve_colors` and it will give you the default color cycle for the specified number of classes. Would you mind trying that?

Reply:
I think resolve_colors func doesn't work as expected.   If both color and colormap are set to None then invoking resolve_colors triggers a TypeError.


```python

def resolve_colors(num_colors=None, colormap=None, color=None):
            colormap = cm.get_cmap(colormap)

    # Work with the colormap
    if color is None and colormap is None:
        if isinstance(colormap, str):
            cmap = colormap
            colormap = cm.get_cmap(colormap)

            if colormap is None:
                raise YellowbrickValueError(
                    "Colormap {0} is not a valid matploblib cmap".format(cmap)
                )

        colors = list(map(colormap, np.linspace(0, 1, num=num_colors)))

    # Work with the color list
    elif color is not None:

        if colormap is not None:
            warnings.warn(
                "'color' and 'colormap' cannot be used simultaneously! Using 'color'."
            )

        colors = list(color) # Ensure colors is a list

    # Get the default colors
    else:
        colors = get_color_cycle()

    if len(colors) != num_colors:
        multiple = num_colors // len(colors) - 1
        mod = num_colors % len(colors)
        colors += multiple * colors                                                                                          colors += colors[:mod]

    return colors
```

Reply:
Whoops, I guess I was wrong about that ... thanks for checking though. Ok, so there is a way to get the default colors and I'm going to have to search for it. One sec ...

Reply:
Ok, after much investigation, it does look like it's supposed to be `resolve_colors` but the function isn't working. I tried to change it but it broke some tests so I'll need more time to get a fix going. I'll work on it and ping you when I do the PR to resolve it. 

In the meantime, perhaps try:

```
# HACK: Waiting for @bbengfort to fix resolve_colors 
color_values = get_color_cycle()
n_colors = len(color_values) 
colors = [color_values[i % n_colors] for i in np.arange(len(indices))]
```

It's ugly, but it should get the job done.

Reply:
👍

Reply:
@bbengfort The default color scheme is leading to information being lost in the heatmap.  This is because the scheme consists of only 6 colors that repeat themselves until all classes have been assigned a color.   Therefore, all datasets with more than 6 classes will have multiple bars with the same primary color thus data loss.

This is why I liked your original color scheme(SPECTRAL) choice; I think it consists of 11 colors.   

How many classes will ClassPredictionError support?   And can we adopt a color scheme that is flexible to larger number of classes while still being visually decipherable ?

Reply:
Yeh, that is a problem - and that's why the user really does need to specify the color for > 6 classes; which is less common but definitely a possibility. I'm not sure you have to do anything about it, but you could pass `spectral` as the name of the colormap for your example. 

Generally, we follow [Cynthia Brewer's advice](http://colorbrewer2.org/#type=sequential&scheme=BuGn&n=3) when it comes to color, so if you'd prefer you could use `Set1` as the default; when I fix `resolve_colors` I do plan on having a color cycle that is a bit longer using this scheme.